linux/libdriver/ioctl/rollup.c

```diff
@@ -0,0 +1,96 @@
+#include <stdio.h>
+#include <unistd.h>
+#include <stdint.h>
+#include <stdint.h>
+#include <string.h>
+#include <fcntl.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <errno.h>
+#include <linux/cartesi/rollup.h>
+#include "rollup.h"
+
+int rollup_driver_init(struct rollup_driver *me)
+{
+	int rc = 0;
+
+	me->fd = open("/dev/rollup", O_RDWR);
+	if (me->fd < 0)
+		return errno;
+
+	struct rollup_setup setup;
+	if ((rc = ioctl(me->fd, IOCTL_ROLLUP_SETUP, &setup))) {
+		rc = errno;
+		goto setup_failed;
+	}
+
+	me->tx->length = setup.tx.length;
+	me->tx->data = mmap((void *)setup.tx.data, setup.tx.length,
+	                    PROT_READ | PROT_WRITE,
+	                    MAP_SHARED | MAP_SHARED, me->fd, 0);
+	if (me->tx->data == MAP_FAILED) {
+		rc = ENOBUFS;
+		goto tx_map_failed;
+	}
+
+	me->rx->length = setup.rx.length;
+	me->rx->data = mmap((void *)setup.rx.data, setup.rx.length,
+	                    PROT_READ,
+	                    MAP_SHARED | MAP_SHARED, me->fd, 0);
+	if (me->rx->data == MAP_FAILED) {
+		rc = ENOBUFS;
+		goto rx_map_failed;
+	}
+	return 0;
+
+rx_map_failed:
+	munmap(me->tx->data, me->tx->length);
+tx_map_failed:
+setup_failed:
+	close(me->fd);
+	return rc;
+}
+
+void rollup_driver_fini(struct rollup_driver *me)
+{
+	munmap(me->tx->data, me->tx->length);
+	munmap(me->rx->data, me->rx->length);
+	close(me->fd);
```

Comment:
It's better to reset all `me` fields uninitializing, in case the user tries to call rollup driver functions again the failure will be more deterministic. A detail is that `fd` must not be set to 0, but to `-1`.